From 27b70565bf45d12af2a263c6bbb420d1df8c983f Mon Sep 17 00:00:00 2001
From: Joseph Guo <qijian.guo@nxp.com>
Date: Wed, 19 Jun 2024 13:12:01 +0300
Subject: [PATCH 06/19] arm64: dts: add imx93-11x11-frdm-tianma-wvga-panel dts

Add Tianma TM050RDH03 5.0" WVGA TFT LCD panel support on i.MX93 FRDM

Signed-off-by: Joseph Guo <qijian.guo@nxp.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |   2 +-
 .../imx93-11x11-frdm-tianma-wvga-panel.dts    | 181 ++++++++++++++++++
 2 files changed, 182 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm64/boot/dts/freescale/imx93-11x11-frdm-tianma-wvga-panel.dts

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 6591a0508709..68f81cdc2308 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -385,7 +385,7 @@ dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-evk.dtb \
 			  imx93-11x11-evk-iw612-otbr.dtb \
 			  imx93-11x11-evk-rpmsg.dtb imx93-11x11-evk-rpmsg-lpv.dtb \
 			  imx93-11x11-frdm.dtb imx93-11x11-frdm-mt9m114.dtb \
-			  imx93-11x11-frdm-dsi.dtb
+			  imx93-11x11-frdm-dsi.dtb imx93-11x11-frdm-tianma-wvga-panel.dtb
 
 dtb-$(CONFIG_ARCH_MXC) += imx91-11x11-evk.dtb \
 			  imx91-11x11-evk-flexspi-m2.dtb imx91-11x11-evk-flexspi-nand-m2.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-tianma-wvga-panel.dts b/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-tianma-wvga-panel.dts
new file mode 100644
index 000000000000..f3e921ae95b5
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-tianma-wvga-panel.dts
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2024 NXP
+ */
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pwm/pwm.h>
+#include "imx93-11x11-frdm.dts"
+
+&lvds_bridge {
+	status = "disabled";
+};
+
+&ldb {
+	status = "disabled";
+};
+
+&ldb_phy {
+	status = "disabled";
+};
+
+/ {
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&tpm5 3 40000 PWM_POLARITY_INVERTED>;	/* 25KHz PWM */
+		brightness-levels = <0 100>;
+		num-interpolated-steps = <100>;
+		default-brightness-level = <80>;
+		power-supply = <&reg_vexp_5v>;
+	};
+
+	panel {
+		compatible = "tianma,tm050rdh03";
+		standby-gpios = <&gpio2 27 GPIO_ACTIVE_LOW>;	/* STBYB */
+		backlight = <&backlight>;
+		power-supply = <&reg_vexp_3v3>;
+
+		port {
+			panel_in: endpoint {
+				remote-endpoint = <&display_out>;
+			};
+		};
+	};
+	reg_vexp_3v3: regulator-vexp-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "VEXP_3V3";
+		gpio = <&pcal6524 2 GPIO_ACTIVE_HIGH>;
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		enable-active-high;
+		vin-supply = <&buck4>;
+	};
+
+	reg_vexp_5v: regulator-vexp-5v {
+		compatible = "regulator-fixed";
+		regulator-name = "VEXP_5V";
+		gpio = <&pcal6524 8 GPIO_ACTIVE_HIGH>;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+	};
+};
+
+&gpio2 {
+	/* avoid power leakage */
+	disp-clk-hog {
+		gpio-hog;
+		gpios = <0 GPIO_ACTIVE_HIGH>;
+		output-low;
+	};
+
+	disp-de-hog {
+		gpio-hog;
+		gpios = <1 GPIO_ACTIVE_HIGH>;
+		output-low;
+	};
+
+	disp-vsync-hog {
+		gpio-hog;
+		gpios = <2 GPIO_ACTIVE_HIGH>;
+		output-low;
+	};
+
+	disp-hsync-hog {
+		gpio-hog;
+		gpios = <3 GPIO_ACTIVE_HIGH>;
+		output-low;
+	};
+};
+
+&iomuxc {
+	pinctrl_tpm5_ch3: tpm5ch3grp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO26__TPM5_CH3	0x31e
+		>;
+	};
+	pinctrl_lcdif: lcdifgrp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO00__MEDIAMIX_DISP_CLK		0x31e
+			MX93_PAD_GPIO_IO01__MEDIAMIX_DISP_DE		0x31e
+			MX93_PAD_GPIO_IO02__MEDIAMIX_DISP_VSYNC		0x31e
+			MX93_PAD_GPIO_IO03__MEDIAMIX_DISP_HSYNC		0x31e
+			MX93_PAD_GPIO_IO04__MEDIAMIX_DISP_DATA00		0x31e
+			MX93_PAD_GPIO_IO05__MEDIAMIX_DISP_DATA01		0x31e
+			MX93_PAD_GPIO_IO06__MEDIAMIX_DISP_DATA02		0x31e
+			MX93_PAD_GPIO_IO07__MEDIAMIX_DISP_DATA03		0x31e
+			MX93_PAD_GPIO_IO08__MEDIAMIX_DISP_DATA04		0x31e
+			MX93_PAD_GPIO_IO09__MEDIAMIX_DISP_DATA05		0x31e
+			MX93_PAD_GPIO_IO10__MEDIAMIX_DISP_DATA06		0x31e
+			MX93_PAD_GPIO_IO11__MEDIAMIX_DISP_DATA07		0x31e
+			MX93_PAD_GPIO_IO12__MEDIAMIX_DISP_DATA08		0x31e
+			MX93_PAD_GPIO_IO13__MEDIAMIX_DISP_DATA09		0x31e
+			MX93_PAD_GPIO_IO14__MEDIAMIX_DISP_DATA10	0x31e
+			MX93_PAD_GPIO_IO15__MEDIAMIX_DISP_DATA11	0x31e
+			MX93_PAD_GPIO_IO16__MEDIAMIX_DISP_DATA12	0x31e
+			MX93_PAD_GPIO_IO17__MEDIAMIX_DISP_DATA13	0x31e
+			MX93_PAD_GPIO_IO18__MEDIAMIX_DISP_DATA14	0x31e
+			MX93_PAD_GPIO_IO19__MEDIAMIX_DISP_DATA15	0x31e
+			MX93_PAD_GPIO_IO20__MEDIAMIX_DISP_DATA16	0x31e
+			MX93_PAD_GPIO_IO21__MEDIAMIX_DISP_DATA17	0x31e
+			MX93_PAD_GPIO_IO27__GPIO2_IO27			0x31e
+		>;
+	};
+	pinctrl_lcdif_gpio: lcdifgpiogrp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO00__GPIO2_IO00			0x51e
+			MX93_PAD_GPIO_IO01__GPIO2_IO01			0x51e
+			MX93_PAD_GPIO_IO02__GPIO2_IO02			0x51e
+			MX93_PAD_GPIO_IO03__GPIO2_IO03			0x51e
+		>;
+	};
+};
+
+&lcdif {
+	assigned-clock-rates = <300000000>, <30000000>, <400000000>, <133333333>;
+	status = "okay";
+};
+
+&parallel_disp_fmt {
+	pinctrl-names = "default", "gpio";
+	pinctrl-0 = <&pinctrl_lcdif>;
+	pinctrl-1 = <&pinctrl_lcdif_gpio>;
+	fsl,interface-pix-fmt = "rgb666";
+	status = "okay";
+
+	ports {
+		port@1 {
+			reg = <1>;
+
+			display_out: endpoint {
+				remote-endpoint = <&panel_in>;
+			};
+		};
+	};
+};
+
+/* pin conflicts */
+&lpspi3 {
+	status = "disabled";
+};
+
+&adp5585_isp {
+	status = "disabled";
+};
+
+&ap1302 {
+	status = "disabled";
+};
+
+&sai3 {
+	status = "disabled";
+};
+
+&flexcan2 {
+	status = "disabled";
+};
+
+&tpm5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_tpm5_ch3>;
+	status = "okay";
+};
-- 
2.25.1

